### DISCUSSION TITLE
Gradle wrapper won't update wrapper JAR and wrapper scripts for certain projects

### DISCUSSION THREAD

**Discussion by davidburstrom** [renovatebot/renovate#26050] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I have not seen this working

### Describe the problem

TLDR: Run `./gradlew :wrapper` instead of `./gradle wrapper`, here's the reproducer: https://github.com/davidburstrom/renovate-gradle-wrapper-bug-reproducer

When Renovate updates the Gradle wrappers, it runs using the `wrapper` tasks (https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gradle-wrapper/artifacts.ts#L149C15-L149C15).

For certain projects, such as https://github.com/davidburstrom/version-compatibility-gradle-plugin, the subprojects are configured to check that the current Gradle version is listed in a set of Gradle versions for compatibility testing. The Gradle configuration fails when Renovate runs, emitting a subtle `WARN: Error executing gradle wrapper update command. It can be not a critical one though.` in the logs, and thus it never updates the wrapper JAR and wrapper scripts.

Renovate really should be running `:wrapper` instead (note the absolute path to the task), which, if the project has turned on [configuration on demand](https://docs.gradle.org/current/userguide/multi_project_configuration_and_execution.html#sec:configuration_on_demand), will only configure the root project.

Obviously, when the PR actually runs in CI it will fail for the same reason, but by that time the relevant Gradle files have been updated at least.
A minor benefit of this change is that the renovate job runs slightly faster too. 

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Executing command (branch="renovate/gradle-8.x")
{
  "command": "docker run --rm --name=renovate_a_sidecar --label=renovate_a_child --memory=3584m -v \"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\":\"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\" -v \"/tmp/worker/66094a/190960/cache\":\"/tmp/worker/66094a/190960/cache\" -e GRADLE_OPTS -e CONTAINERBASE_CACHE_DIR -w \"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\" ghcr.io/containerbase/sidecar:9.24.0 bash -l -c \"install-tool java 17.0.9+9 && ./gradlew wrapper --gradle-distribution-url https://services.gradle.org/distributions/gradle-8.5-bin.zip\""
}

DEBUG: rawExec err (branch="renovate/gradle-8.x")
{
  "err": {
    "cmd": "/bin/sh -c docker run --rm --name=renovate_a_sidecar --label=renovate_a_child --memory=3584m -v \"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\":\"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\" -v \"/tmp/worker/66094a/190960/cache\":\"/tmp/worker/66094a/190960/cache\" -e GRADLE_OPTS -e CONTAINERBASE_CACHE_DIR -w \"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\" ghcr.io/containerbase/sidecar:9.24.0 bash -l -c \"install-tool java 17.0.9+9 && ./gradlew wrapper --gradle-distribution-url https://services.gradle.org/distributions/gradle-8.5-bin.zip\"",
    "stderr": "\nFAILURE: Build failed with an exception.\n\n* Where:\nBuild file '/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer/subproject/build.gradle.kts' line: 2\n\n* What went wrong:\nplease update stuff\n\n* Try:\n> Run with --stacktrace option to get the stack trace.\n> Run with --info or --debug option to get more log output.\n> Run with --scan to get full insights.\n> Get more help at https://help.gradle.org.\n\nBUILD FAILED in 31s\n",
    "stdout": "installing v2 tool java v17.0.9+9\n[22:17:15.659] INFO (9): Installing tool java v17.0.9+9...\nlinking tool java v17.0.9+9\nopenjdk version \"17.0.9\" 2023-10-17\nOpenJDK Runtime Environment Temurin-17.0.9+9 (build 17.0.9+9)\nOpenJDK 64-Bit Server VM Temurin-17.0.9+9 (build 17.0.9+9, mixed mode, sharing)\n[22:17:19.214] INFO (9): Installed tool java in 3.5s.\nDownloading https://services.gradle.org/distributions/gradle-8.5-bin.zip\n............10%.............20%............30%.............40%.............50%............60%.............70%.............80%............90%.............100%\n\nWelcome to Gradle 8.5!\n\nHere are the highlights of this release:\n - Support for running on Java 21\n - Faster first use with Kotlin DSL\n - Improved error and warning messages\n\nFor more details see https://docs.gradle.org/8.5/release-notes.html\n\nTo honour the JVM settings for this build a single-use Daemon process will be forked. For more on this, please refer to https://docs.gradle.org/8.5/userguide/gradle_daemon.html#sec:disabling_the_daemon in the Gradle documentation.\nDaemon will be stopped at the end of the build \nConfiguration on demand is an incubating feature.\n",
    "options": {
      "cwd": "/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer",
      "encoding": "utf-8",
      "env": {
        "GRADLE_OPTS": "-Dorg.gradle.parallel=true -Dorg.gradle.configureondemand=true -Dorg.gradle.daemon=false -Dorg.gradle.caching=false",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/.cargo/bin:/home/ubuntu/.local/bin:/go/bin:/home/ubuntu/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "CONTAINERBASE_CACHE_DIR": "/tmp/worker/66094a/190960/cache/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "name": "ExecError",
    "message": "Command failed: docker run --rm --name=renovate_a_sidecar --label=renovate_a_child --memory=3584m -v \"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\":\"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\" -v \"/tmp/worker/66094a/190960/cache\":\"/tmp/worker/66094a/190960/cache\" -e GRADLE_OPTS -e CONTAINERBASE_CACHE_DIR -w \"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\" ghcr.io/containerbase/sidecar:9.24.0 bash -l -c \"install-tool java 17.0.9+9 && ./gradlew wrapper --gradle-distribution-url https://services.gradle.org/distributions/gradle-8.5-bin.zip\"\n\nFAILURE: Build failed with an exception.\n\n* Where:\nBuild file '/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer/subproject/build.gradle.kts' line: 2\n\n* What went wrong:\nplease update stuff\n\n* Try:\n> Run with --stacktrace option to get the stack trace.\n> Run with --info or --debug option to get more log output.\n> Run with --scan to get full insights.\n> Get more help at https://help.gradle.org.\n\nBUILD FAILED in 31s\n",
    "stack": "ExecError: Command failed: docker run --rm --name=renovate_a_sidecar --label=renovate_a_child --memory=3584m -v \"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\":\"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\" -v \"/tmp/worker/66094a/190960/cache\":\"/tmp/worker/66094a/190960/cache\" -e GRADLE_OPTS -e CONTAINERBASE_CACHE_DIR -w \"/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer\" ghcr.io/containerbase/sidecar:9.24.0 bash -l -c \"install-tool java 17.0.9+9 && ./gradlew wrapper --gradle-distribution-url https://services.gradle.org/distributions/gradle-8.5-bin.zip\"\n\nFAILURE: Build failed with an exception.\n\n* Where:\nBuild file '/tmp/worker/66094a/190960/repos/github/davidburstrom/renovate-gradle-wrapper-bug-reproducer/subproject/build.gradle.kts' line: 2\n\n* What went wrong:\nplease update stuff\n\n* Try:\n> Run with --stacktrace option to get the stack trace.\n> Run with --info or --debug option to get more log output.\n> Run with --scan to get full insights.\n> Get more help at https://help.gradle.org.\n\nBUILD FAILED in 31s\n\n    at ChildProcess.<anonymous> (/opt/containerbase/tools/renovate/37.59.8/node_modules/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:529:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:292:12)"
  }
  "durationMs": 37037
}

WARN: Error executing gradle wrapper update command. It can be not a critical one though. (branch="renovate/gradle-8.x")```

</details>


### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
Thanks @davidburstrom 

Do I understand right that Renovate should _always_ run this way, for all projects?

**Reply by davidburstrom** [reply]
@rarkins Yes, that's the proposal. It would lead to a slight change of behaviour, but I can only think of contrived cases where this would be a regression (for example if a project has subprojects that have defined custom tasks that are also named `'wrapper'` and it relies on `./gradlew wrapper` instead of `./gradlew :wrapper` to execute those. If there ever was such a project, it can be reconfigured so that the wrapper task is either finalized by or depends on the custom tasks). From the point of Renovate, the Gradle version bump would be more correct, as the jar and scripts will be updated in the PR.

This change would not affect [composed builds](https://docs.gradle.org/current/userguide/composite_builds.html), because the wrapper in the included builds is not updated by running `./gradlew wrapper`. I've created [another plugin that does that](https://github.com/davidburstrom/recursive-wrapper-gradle-plugin).

**Reply by rarkins** [reply]
@Churro wdyt?

**Reply by Churro** [reply]
@davidburstrom, wouldn't applying the `wrapper` task only on the root project mean that breaking gradle changes between versions are not surfaced during wrapper updates if they affect subprojects? I'm assuming the use of a single wrapper here, not using different ones in subprojects (in that case, using composed builds make more sense, of course).

I currently can't think of another reason why all gradle docs suggest exclusively using `./gradlew wrapper` and not `./gradlew :wrapper`, e.g. [here](https://docs.gradle.org/8.5/release-notes.html). Would be great if you could share your thoughts on that.

**Reply by davidburstrom** [reply]
@Churro That is correct, breaking changes won't be surfaced. But that's already the case in Renovate: Renovate will update the versions in the properties file and open a PR that won't build in CI. What my proposed change will do differently is that it will increase the likelihood that Renovate will be able to update the wrapper JAR and scripts.

As for `./gradlew wrapper` vs `./gradlew :wrapper`, it's a good question. If you're using the Wrapper task to update the Gradle version, you actually need to run the wrapper task twice: First time it runs in the current version of Gradle to update the properties (and it won't update the JAR or scripts). During that first run, it's thus pointless to configure the entire project. Second time the new version of Gradle will be provided and run, which possibly writes the new JAR and scripts. A lot of people don't know that a second run is required, which means that they will discover the breaking changes next time they build, and probably won't update the wrapper JAR and scripts properly. From that reasoning, I cannot think that the docs indicate `wrapper` instead of `:wrapper` in order to surface incompatibilities, because it should also indicate that it should be run twice. I've never seen a case for running `wrapper` instead of `:wrapper` but that obviously doesn't mean that a case doesn't exist.

There's another approach that could be pursued too, which is inspired by https://blog.marcnuri.com/gradle-wrapper-from-docker: Run the wrapper task in Docker to generate the JAR and scripts and copy them out of the container into the repo. But such an approach wouldn't update the wrappers in a composite build (if for example the recursive wrapper Gradle plugin is being used).

I should also add that if the Renovate behaviour is changed, it's also reversible in case it creates more problems than it solves.

**Reply by Churro** [reply, marked as answer]
@davidburstrom, thanks.

I gave it a try with several projects, some containing sub-projects, one requiring a lock file refresh afterwards, etc and didn't notice anything breaking. Still, my sample set is nothing compared to all the variations of projects renovate is applied to. PR https://github.com/renovatebot/renovate/pull/26098 implements the change and highlights its effects. Of course, it can be reverted but in general this should be the last resort. 

`gradle-wrapper` updates are comparably slow, so I'm -slightly- optimistic that this will have a positive on renovate runtimes overall. On a bigger project in direct comparison, it was a delta of 56 sec before and after the PR, so overall this should have a noticeable positive effect on renovate runtimes.

**Reply by davidburstrom** [reply]
Nice, thank you for implementing this! I look forward to use the release!
